homeassistant/components/usb/__init__.py

```diff
@@ -0,0 +1,137 @@
+from homeassistant.const import EVENT_HOMEASSISTANT_STARTED, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.loader import async_get_usb
+
+from .flow import FlowDispatcher, USBFlow
+from .models import USBDevice
+from .utils import USBDeviceTupleType, usb_device_from_port, usb_device_tuple
+
+_LOGGER = logging.getLogger(__name__)
+
+# Perodic scanning only happens on non-linux systems
+SCAN_INTERVAL = datetime.timedelta(minutes=60)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the USB Discovery integration."""
+    usb = await async_get_usb(hass)
+    usb_discovery = USBDiscovery(hass, FlowDispatcher(hass), usb)
+    await usb_discovery.async_setup()
+    return True
+
+
+class USBDiscovery:
+    """Manage USB Discovery."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        flow_dispatcher: FlowDispatcher,
+        usb: list[dict[str, str]],
+    ) -> None:
+        """Init USB Discovery."""
+        self.hass = hass
+        self.flow_dispatcher = flow_dispatcher
+        self.usb = usb
+        self.seen: set[USBDeviceTupleType] = set()
+
+    async def async_setup(self) -> None:
+        """Set up USB Discovery."""
+        if not await self._async_start_monitor():
+            await self._async_start_scanner()
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, self.async_start)
+
+    async def async_start(self, event: Event) -> None:
+        """Start USB Discovery and run a manual scan."""
+        self.flow_dispatcher.async_start()
+        await self.hass.async_add_executor_job(self.scan_serial)
+
+    async def _async_start_scanner(self) -> None:
+        """Perodic scan with pyserial when the observer is not available."""
+        stop_track = async_track_time_interval(
+            self.hass, lambda now: self.scan_serial(), SCAN_INTERVAL
+        )
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, callback(lambda event: stop_track())
+        )
+
+    async def _async_start_monitor(self) -> bool:
+        """Start monitoring hardware with pyudev."""
+        if not sys.platform.startswith("linux"):
+            return False
+        from pyudev import (  # pylint: disable=import-outside-toplevel
+            Context,
+            Monitor,
+            MonitorObserver,
+        )
+
+        try:
+            context = Context()
+        except ImportError:
+            return False
+
+        monitor = Monitor.from_netlink(context)
+        monitor.filter_by(subsystem="tty")
+        observer = MonitorObserver(
+            monitor, callback=self._device_discovered, name="usb-observer"
+        )
+        observer.start()
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, lambda event: observer.stop()
```

Comment:
`observer.stop` will run in the thread pool like this.

Reply:
Yes, its joining the thread so it could block